synfig-studio/src/gui/asyncrenderer.h

```diff
@@ -102,6 +110,7 @@ class AsyncRenderer : public etl::shared_object, public sigc::trackable
 
 	void start();
 	void stop();
+	void user_stop();
```

Comment:
No need for it: just change previous line `stop()` to `stop(Interaction interaction = UNDEFINED)`.
When it's a user interaction, call `stop(BY_USER)` ;)

Reply:
Or `stop(AsyncRender::BY_USER)` if there's a namespace problem.

Reply:
I cannot get this to work...

Reply:
What's happening?

Reply:
When I define `Interaction` with it's default `UNDEFINED` value, the compiler gives an error about the member function definition...
I don't have the code on this machine, but I suspect it's related to calling a member variable of a class within its method, later I'll report here the compiler error.

Reply:
Here you can find my compilation output: https://pastebin.com/xjaBNZL7

Reply:
Here you can find the changes applied with the previous output: https://pastebin.com/XKAaZuUi

Reply:
Hm… Apparently sigc connections don't apply the default value for callback arguments :p

Two possible solutions IMO:
- adopt the java overloading style: two methods: stop() and stop(Interaction) methods. The former calls the last filling the argument with default value.
- make use of sigc::bind, as described here: https://developer.gnome.org/gtkmm-tutorial/stable/sec-binding-extra-arguments.html.en https://developer.gnome.org/libsigc++-tutorial/stable/ch04.html https://developer.gnome.org/gtkmm-tutorial/stable/sec-connecting-signal-handlers.html.en )

Reply:
mmm, I should use `sigc::bind(function, value_to_pass)` right?

Tried with this:
`stop_button.signal_clicked().connect(sigc::bind(sigc::mem_fun(*this, &studio::Dock_Info::on_stop_button_clicked)), studio::AsyncRenderer::INTERACTION_BY_USER);`

but still getting compiling error... honestly no idea why...

Reply:
My diff so far: https://pastebin.com/kmWw02BP

Reply:
You would bind a AsyncRenderer::stop() call, not the callback ;)
Anyway, the problem isn't at this line.

In your previous compilation error report, you can (hardly) find that the problem in in line 922 of asyncrenderer.cpp (on some version of it…).

![Captura de tela de 2020-08-07 19-54-16](https://user-images.githubusercontent.com/1323987/89694673-d1600f80-d8e7-11ea-822d-9979ebf3b896.png)

Here's my tip for those cases GCC spills out a lot of strange error messages. ;) select the whole error block, copy an pasted in a text editor and search for path or namespaces it's exclusively in your code (not in those #included or linked): I used "gui/" on this case. "required from here" is good too. With regex it's easier to filter out what of your source code is triggering this mess.
Then you look at the line/column it describes and... have hope lol

```
#ifdef GLIB_DISPATCHER_BROKEN
		done_connection=Glib::signal_timeout().connect(
			sigc::bind_return(
				mem_fun(*this,&AsyncRenderer::stop),   // HERE
				false
			)
			,0
		);
```

The AsynRenderer::stop() signature changed to AsynRenderer::stop(enum), right? So that call misses the argument now.

Sorry I didn't guide you properly to the problematic line before

Reply:
No problems, you're very helpful 👍 
I found that error message (console output colors errors and warnings) but the error message itself does not give information because it indicates the last line of the edited file.

btw gnome documentantion isn't very easy to follow, at least for me it's missing lots of stuff...

I should bind directly the method like this as reference:
`sigc::bind(async_renderer.stop(), studio::AsyncRenderer::INTERACTION_BY_USER);`

Regaring the example you pointed out, I should use `bind()` instead of `bind_return()`, right? This way I bind the "input" values.

I looked at some other examples in the source code of Synfig and I'm starting to get a clearer idea of what should be done.

I'll try later to apply those changes and report here my updates.

Reply:
Followed your tips and now compiles fine :smile:
Still needs some tuning though.

Reply:
Here you find the changes: https://pastebin.com/YSDwk240

At this point compilation is fine, button is enabled, BUT rendering isn't stopped after confirmation.

Reply:
This is the reason:
@@ -213,7 +214,10 @@ void studio::Dock_Info::on_stop_button_clicked()
        Gtk::MESSAGE_QUESTION,
        _("Cancel"),
        _("Stop")))
-       async_renderer->user_stop();
+       sigc::bind(
+           sigc::mem_fun(*async_renderer, &studio::AsyncRenderer::stop),
+           studio::AsyncRenderer::INTERACTION_BY_USER
+       );
 }

neither `sigc::bind()` or `sigc::mem_fun()` call a method: they just a way to wrap or to point to a class method like it was a C-style function (i.e., make a class method be called as it was a regular function). ;)

Reply:
I've seen many samples in the source code where the binded method is connected to a signal, like this:
`menu->signal_hide().connect(sigc::bind(sigc::ptr_fun(&delete_widget), menu));`

but in our case `stop()` does not have a `signal_something()` to `connect()` to and call the method...

I don't have experience with this stuff, how would you do that?

Reply:
Exactly. It's meant to be used for signal connections (slots): what function should signal calls when it's emitted. The problem is that in C++, we usually want to call an object method, that depends on... the object ;) 
Gtkmm use a similar approach as Qt (now) does (https://wiki.qt.io/New_Signal_Slot_Syntax)

Anyway,  we just need here a regular call.
`stop(INTERACTION_BY_USER)` :)

Reply:
Finally done :yum: